File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/bridge/JavaOnlyArray.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react.bridge;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Java {@link ArrayList} backed implementation of {@link ReadableArray} and {@link WritableArray}
 * Instances of this class SHOULD NOT be used for communication between java and JS, use instances
 * of {@link WritableNativeArray} created via {@link Arguments#createArray} or just {@link
 * ReadableArray} interface if you want your "native" module method to take an array from JS as an
 * argument.
 *
 * <p>Main purpose for this class is to be used in java-only unit tests, but could also be used
 * outside of tests in the code that operates only in java and needs to communicate with RN modules
 * via their JS-exposed API.
 */
public class JavaOnlyArray implements ReadableArray, WritableArray {

  private final List mBackingList;

  public static JavaOnlyArray from(List list) {
    return new JavaOnlyArray(list);
  }

  public static JavaOnlyArray of(Object... values) {
    return new JavaOnlyArray(values);
  }

  public static JavaOnlyArray deepClone(ReadableArray ary) {
    JavaOnlyArray res = new JavaOnlyArray();
    for (int i = 0, size = ary.size(); i < size; i++) {
      ReadableType type = ary.getType(i);
      switch (type) {
        case Null:
          res.pushNull();
          break;
        case Boolean:
          res.pushBoolean(ary.getBoolean(i));
          break;
        case Number:
          res.pushDouble(ary.getDouble(i));
          break;
        case String:
          res.pushString(ary.getString(i));
          break;
        case Map:
          res.pushMap(JavaOnlyMap.deepClone(ary.getMap(i)));
          break;
        case Array:
          res.pushArray(deepClone(ary.getArray(i)));
          break;
      }
    }
    return res;
  }

  private JavaOnlyArray(Object... values) {
    mBackingList = Arrays.asList(values);
  }

  private JavaOnlyArray(List list) {
    mBackingList = new ArrayList(list);
  }

  public JavaOnlyArray() {
    mBackingList = new ArrayList();
  }

  @Override
  public int size() {
    return mBackingList.size();
  }

  @Override
  public boolean isNull(int index) {
    return mBackingList.get(index) == null;
  }

  @Override
  public double getDouble(int index) {
    return ((Number) mBackingList.get(index)).doubleValue();
  }

  @Override
  public int getInt(int index) {
    return ((Number) mBackingList.get(index)).intValue();
  }

  @Override
  public @Nullable String getString(int index) {
    return (String) mBackingList.get(index);
  }

  @Override
  public ReadableArray getArray(int index) {
    return (ReadableArray) mBackingList.get(index);
  }

  @Override
  public boolean getBoolean(int index) {
    return (Boolean) mBackingList.get(index);
  }

  @Override
  public ReadableMap getMap(int index) {
    return (ReadableMap) mBackingList.get(index);
  }

  @Override
  public @NonNull Dynamic getDynamic(int index) {
    return DynamicFromArray.create(this, index);
  }

  @Override
  public @NonNull ReadableType getType(int index) {
    Object object = mBackingList.get(index);

    if (object == null) {
      return ReadableType.Null;
    } else if (object instanceof Boolean) {
      return ReadableType.Boolean;
    } else if (object instanceof Double || object instanceof Float || object instanceof Integer) {
      return ReadableType.Number;
    } else if (object instanceof String) {
      return ReadableType.String;
    } else if (object instanceof ReadableArray) {
      return ReadableType.Array;
    } else if (object instanceof ReadableMap) {
      return ReadableType.Map;
    }
    return null;
  }

  @Override
  public void pushBoolean(boolean value) {
    mBackingList.add(value);
  }

  @Override
  public void pushDouble(double value) {
    mBackingList.add(value);
  }

  @Override
  public void pushInt(int value) {
    mBackingList.add(new Double(value));
  }

  @Override
  public void pushString(@Nullable String value) {
    mBackingList.add(value);
  }

  @Override
  public void pushArray(@Nullable ReadableArray array) {
    mBackingList.add(array);
  }

  @Override
  public void pushMap(@Nullable ReadableMap map) {
    mBackingList.add(map);
  }

  @Override
  public void pushNull() {
    mBackingList.add(null);
  }

  @Override
  public @NonNull ArrayList<Object> toArrayList() {
    return new ArrayList<Object>(mBackingList);
  }

  @Override
  public String toString() {
    return mBackingList.toString();
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;

    JavaOnlyArray that = (JavaOnlyArray) o;

    if (mBackingList != null ? !mBackingList.equals(that.mBackingList) : that.mBackingList != null)
      return false;

    return true;
  }

  @Override
  public int hashCode() {
    return mBackingList != null ? mBackingList.hashCode() : 0;
  }
}

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/bridge/JavaOnlyArray.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react.bridge

import com.facebook.react.bridge.JavaOnlyMap.Companion.deepClone
import java.util.ArrayList

/**
 * Java [ArrayList] backed implementation of [ReadableArray] and [WritableArray] Instances of this
 * class SHOULD NOT be used for communication between java and JS, use instances of
 * [WritableNativeArray] created via [Arguments#createArray] or just [ReadableArray] interface if
 * you want your "native" module method to take an array from JS as an argument.
 *
 * Main purpose for this class is to be used in java-only unit tests, but could also be used outside
 * of tests in the code that operates only in java and needs to communicate with RN modules via
 * their JS-exposed API.
 */
public class JavaOnlyArray : ReadableArray, WritableArray {

  public companion object {
    @JvmStatic public fun from(list: List<*>): JavaOnlyArray = JavaOnlyArray(list)

    @JvmStatic public fun of(vararg values: Any?): JavaOnlyArray = JavaOnlyArray(*values)

    @JvmStatic
    public fun deepClone(array: ReadableArray?): JavaOnlyArray {
      val res = JavaOnlyArray()
      if (array == null) {
        return res
      }

      repeat(array.size()) { i ->
        val type = array.getType(i)
        when (type) {
          ReadableType.Null -> res.pushNull()
          ReadableType.Boolean -> res.pushBoolean(array.getBoolean(i))
          ReadableType.Number -> res.pushDouble(array.getDouble(i))
          ReadableType.String -> res.pushString(array.getString(i))
          ReadableType.Map -> res.pushMap(deepClone(array.getMap(i)))
          ReadableType.Array -> res.pushArray(deepClone(array.getArray(i)))
        }
      }
      return res
    }
  }

  private val backingList: MutableList<Any?>

  private constructor(vararg values: Any?) {
    backingList = mutableListOf(*values)
  }

  private constructor(list: List<*>) {
    backingList = ArrayList(list)
  }

  public constructor() {
    backingList = mutableListOf()
  }

  override fun size(): Int = backingList.size

  override fun isNull(index: Int): Boolean = backingList[index] == null

  override fun getDouble(index: Int): Double = (backingList[index] as Number).toDouble()

  override fun getInt(index: Int): Int = (backingList[index] as Number).toInt()

  override fun getLong(index: Int): Long = (backingList[index] as Number).toLong()

  override fun getString(index: Int): String? = backingList[index] as String?

  override fun getArray(index: Int): ReadableArray? = backingList[index] as ReadableArray?

  override fun getBoolean(index: Int): Boolean = backingList[index] as Boolean

  override fun getMap(index: Int): ReadableMap? = backingList[index] as ReadableMap?

  override fun getDynamic(index: Int): Dynamic = DynamicFromArray.create(this, index)

  override fun getType(index: Int): ReadableType {
    return when (val value = backingList[index]) {
      null -> ReadableType.Null
      is Boolean -> ReadableType.Boolean
      is Double,
      is Float,
      is Int,
      is Long -> ReadableType.Number
      is String -> ReadableType.String
      is ReadableArray -> ReadableType.Array
      is ReadableMap -> ReadableType.Map
      else -> throw IllegalStateException("Invalid type ${value.javaClass})")
    }
  }

  override fun pushBoolean(value: Boolean) {
    backingList.add(value)
  }

  override fun pushDouble(value: Double) {
    backingList.add(value)
  }

  override fun pushInt(value: Int) {
    backingList.add(value.toDouble())
  }

  override fun pushLong(value: Long) {
    backingList.add(value.toDouble())
  }

  override fun pushString(value: String?) {
    backingList.add(value)
  }

  override fun pushArray(array: ReadableArray?) {
    backingList.add(array)
  }

  override fun pushMap(map: ReadableMap?) {
    backingList.add(map)
  }

  override fun pushNull() {
    backingList.add(null)
  }

  override fun toArrayList(): ArrayList<Any?> = ArrayList(backingList)

  override fun toString(): String = backingList.toString()

  override fun equals(other: Any?): Boolean =
      if (this === other) {
        true
      } else if (other == null || javaClass != other.javaClass) {
        false
      } else {
        backingList == (other as JavaOnlyArray).backingList
      }

  override fun hashCode(): Int = backingList.hashCode()
}
